File: implementation/scipamato/scipamato-core-persistence-jooq/src/test/java/ch/difty/scipamato/core/persistence/code/JooqCodeRepoIntegrationTest.java
package ch.difty.scipamato.core.persistence.code;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.fail;

import java.util.Collection;
import java.util.List;

import lombok.extern.slf4j.Slf4j;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;

import ch.difty.scipamato.common.entity.CodeClassId;
import ch.difty.scipamato.common.persistence.paging.PaginationRequest;
import ch.difty.scipamato.common.persistence.paging.Sort;
import ch.difty.scipamato.core.entity.Code;
import ch.difty.scipamato.core.entity.CodeClass;
import ch.difty.scipamato.core.entity.code.CodeDefinition;
import ch.difty.scipamato.core.entity.code.CodeFilter;
import ch.difty.scipamato.core.entity.code.CodeTranslation;
import ch.difty.scipamato.core.persistence.JooqTransactionalIntegrationTest;
import ch.difty.scipamato.core.persistence.OptimisticLockingException;

@Slf4j
public class JooqCodeRepoIntegrationTest extends JooqTransactionalIntegrationTest {

    @Autowired
    private JooqCodeRepo repo;

    @Test
    public void findingAllCodes1InGerman() {
        List<Code> codesOfClass1 = repo.findCodesOfClass(CodeClassId.CC1, "de");
        assertThat(codesOfClass1)
            .isNotEmpty()
            .hasSize(21);
        codesOfClass1.forEach((c) -> log.debug(c.toString()));
    }

    @Test
    public void findingAllCodes2InEnglish() {
        List<Code> codesOfClass1 = repo.findCodesOfClass(CodeClassId.CC2, "en");
        assertThat(codesOfClass1)
            .isNotEmpty()
            .hasSize(2);
        codesOfClass1.forEach((c) -> log.debug(c.toString()));
    }

    @Test
    public void findingAllCodes3InEnglish() {
        List<Code> codesOfClass1 = repo.findCodesOfClass(CodeClassId.CC3, "fr");
        assertThat(codesOfClass1)
            .isNotEmpty()
            .hasSize(14);
        codesOfClass1.forEach((c) -> log.debug(c.toString()));
    }

    @Test
    public void findingCodeDefinitions_withUnspecifiedFilter_findsAllDefinitions() {
        final List<CodeDefinition> cds = repo.findPageOfCodeDefinitions(new CodeFilter(),
            new PaginationRequest(0, 10, Sort.Direction.ASC, "name"));

        assertThat(cds).hasSize(10);

        CodeDefinition cd = cds.get(0);
        assertThat(cd.getCode()).isEqualTo("4Y");
        assertThat(cd.getCodeClass())
            .extracting("id")
            .containsExactly(4);
        assertThat(cd.getName()).isEqualTo("Absenzen, eingeschränkte Aktivität");
        assertThat(cd.getSort()).isEqualTo(17);
        assertThat(cd.isInternal()).isTrue();
        assertThat(cd.getNameInLanguage("de")).startsWith("Absenzen,");
        assertThat(cd.getNameInLanguage("en")).startsWith("Absenteeism");
        assertThat(cd.getNameInLanguage("fr")).startsWith("Absentéisme");

        cd = cds.get(1);
        assertThat(cd.getCode()).isEqualTo("7Z");
        assertThat(cd.getCodeClass())
            .extracting("id")
            .containsExactly(7);
        assertThat(cd.getName()).isEqualTo("Alle");
        assertThat(cd.getSort()).isEqualTo(4);
        assertThat(cd.isInternal()).isTrue();
        assertThat(cd.getNameInLanguage("de")).startsWith("Alle");
        assertThat(cd.getNameInLanguage("en")).startsWith("All");
        assertThat(cd.getNameInLanguage("fr")).startsWith("tous");
    }

    @Test
    public void findingCodeDefinitions_withFilterMatchingSingleGermanName_findsOne() {
        final CodeFilter filter = new CodeFilter();
        filter.setNameMask("Experimentelle Studie unter Belastung / Arbeit");
        final List<CodeDefinition> kds = repo.findPageOfCodeDefinitions(filter,
            new PaginationRequest(Sort.Direction.ASC, "name"));

        assertThat(kds).hasSize(1);

        CodeDefinition ntd = kds.get(0);
        assertThat(ntd.getCode()).isEqualTo("5A");
        assertThat(ntd.getName()).isEqualTo("Experimentelle Studie unter Belastung / Arbeit");
        assertThat(ntd.getNameInLanguage("de")).isEqualTo("Experimentelle Studie unter Belastung / Arbeit");
        assertThat(ntd.getNameInLanguage("en")).isEqualTo("Experimental study under exercising conditions");
        assertThat(ntd.getNameInLanguage("fr")).isEqualTo("Etude expérimentale dans des conditions exercicantes");
    }

    @Test
    public void findingCodeDefinitions_haveVersionFieldsPopulated() {
        final CodeFilter filter = new CodeFilter();
        filter.setNameMask("Experimentelle Studie unter Belastung / Arbeit");
        final List<CodeDefinition> kds = repo.findPageOfCodeDefinitions(filter,
            new PaginationRequest(Sort.Direction.ASC, "name"));

        assertThat(kds).hasSize(1);

        CodeDefinition cd = kds.get(0);

        assertThat(cd.getVersion()).isEqualTo(1);
        assertThat(cd.getCreated()).isNull();
        assertThat(cd.getLastModified()).isNull();

        Collection<CodeTranslation> translations = cd
            .getTranslations()
            .values();
        assertThat(translations).isNotEmpty();
        CodeTranslation tr = translations
            .iterator()
            .next();
        assertThat(tr.getVersion()).isEqualTo(1);
        assertThat(tr.getCreated()).isNull();
        assertThat(tr.getLastModified()).isNull();
    }

    @Test
    public void countingCodes_withUnspecifiedFilter_findsAllDefinitions() {
        assertThat(repo.countByFilter(new CodeFilter())).isEqualTo(82);
    }

    @Test
    public void countingCodes_withFilter_findsAllMatchingDefinitions() {
        final CodeFilter filter = new CodeFilter();
        filter.setNameMask("es");
        assertThat(repo.countByFilter(filter)).isEqualTo(44);
    }

    @Test
    public void countingCodes_withNonMatchingFilter_findsNone() {
        final CodeFilter filter = new CodeFilter();
        filter.setNameMask("foobar");
        assertThat(repo.countByFilter(filter)).isEqualTo(0);
    }

    @Test
    public void gettingMainLanguage() {
        assertThat(repo.getMainLanguage()).isEqualTo("de");
    }

    @Test
    public void findingMainLanguage() {
        CodeDefinition cd = repo.newUnpersistedCodeDefinition();

        assertThat(cd.getCode()).isNull();
        assertThat(cd.getMainLanguageCode()).isEqualTo("de");
        assertThat(cd.getCodeClass()).isNull();
        assertThat(cd.getSort()).isEqualTo(1);
        assertThat(cd.isInternal()).isFalse();
        assertThat(cd.getName()).isEqualTo("n.a.");
        assertThat(cd.getNameInLanguage("de")).isNull();
        assertThat(cd
            .getTranslations()
            .asMap()).hasSize(3);

        final Collection<CodeTranslation> translations = cd
            .getTranslations()
            .values();
        assertThat(translations)
            .extracting("langCode")
            .containsOnly("de", "en", "fr");
        assertThat(translations)
            .extracting("id")
            .containsExactly(null, null, null);
        assertThat(translations)
            .extracting("name")
            .containsExactly(null, null, null);
    }

    @Test
    public void findingCodeDefinition_withNonExistingId_returnsNull() {
        assertThat(repo.findCodeDefinition("foo")).isNull();
    }

    @Test
    public void findingCodeDefinition_withExistingId_loadsWithAllLanguages() {
        final CodeDefinition existing = repo.findCodeDefinition("4Y");

        assertThat(existing).isNotNull();
        assertThat(existing.getCode()).isEqualTo("4Y");
        assertThat(existing.getName()).startsWith("Absenzen");
        assertThat(existing
            .getTranslations()
            .asMap()).hasSize(3);
        assertThat(existing.getNameInLanguage("de")).startsWith("Absenzen,");
        assertThat(existing.getNameInLanguage("en")).startsWith("Absenteeism");
        assertThat(existing.getNameInLanguage("fr")).startsWith("Absentéisme");
    }

    @Test
    public void savingNewRecord_savesRecordAndRefreshesId() {
        final CodeTranslation ct_de = new CodeTranslation(null, "de", "foo_de", "Kommentar", 0);
        final CodeTranslation ct_en = new CodeTranslation(null, "en", "foo1_en", null, 0);
        final CodeTranslation ct_fr = new CodeTranslation(null, "fr", "foo1_fr", null, 0);
        CodeClass cc = new CodeClass(2, null, null);
        final CodeDefinition cd = new CodeDefinition("2Z", "de", cc, 0, false, 1, ct_de, ct_en, ct_fr);

        assertThat(cd
            .getTranslations()
            .values())
            .extracting("id")
            .containsExactly(null, null, null);

        CodeDefinition saved = repo.saveOrUpdate(cd);

        assertThat(saved).isNotNull();
        assertThat(saved.getCode()).isEqualTo("2Z");
        assertThat(saved.getName()).isEqualTo("foo_de");
        assertThat(saved
            .getTranslations()
            .size()).isEqualTo(3);
        assertThat(saved
            .getTranslations()
            .values())
            .extracting("version")
            .containsExactly(1, 1, 1);
    }

    @Test
    public void updatingRecord() {
        final CodeDefinition cd = repo.findCodeDefinition("2R");

        assertThat(cd).isNotNull();
        assertThat(cd.getCode()).isEqualTo("2R");
        assertThat(cd.getName()).isEqualTo("Europa");
        assertThat(cd
            .getTranslations()
            .asMap()).hasSize(3);
        assertThat(cd.getNameInLanguage("de")).isEqualTo("Europa");
        assertThat(cd.getNameInLanguage("en")).isEqualTo("Europe");
        assertThat(cd.getNameInLanguage("fr")).isEqualTo("Europe");
        assertThat(cd
            .getTranslations()
            .get("de")
            .get(0)
            .getVersion()).isEqualTo(1);
        assertThat(cd
            .getTranslations()
            .get("en")
            .get(0)
            .getVersion()).isEqualTo(1);

        cd.setNameInLanguage("de", "eu");
        cd.setNameInLanguage("fr", "foo");

        CodeDefinition updated = repo.saveOrUpdate(cd);

        assertThat(updated).isNotNull();
        assertThat(updated.getCode()).isEqualTo("2R");
        assertThat(updated
            .getTranslations()
            .asMap()).hasSize(3);
        assertThat(updated.getNameInLanguage("de")).isEqualTo("eu");
        assertThat(updated.getNameInLanguage("en")).isEqualTo("Europe");
        assertThat(updated.getNameInLanguage("fr")).isEqualTo("foo");

        assertThat(updated.getVersion()).isEqualTo(2);
        assertThat(updated
            .getTranslations()
            .get("de")
            .get(0)
            .getVersion()).isEqualTo(2);
        assertThat(updated
            .getTranslations()
            .get("en")
            .get(0)
            .getVersion()).isEqualTo(2);
        assertThat(updated
            .getTranslations()
            .get("fr")
            .get(0)
            .getVersion()).isEqualTo(2);
    }

    @Test
    public void deleting_withNonExistingId_returnsNull() {
        assertThat(repo.delete("ZZ", 1)).isNull();
    }

    @Test
    public void deleting_withExistingId_butWrongVersion_throwsOptimisticLockingException() {
        try {
            repo.delete("1A", -1);
            fail("should have thrown exception");
        } catch (Exception ex) {
            assertThat(ex)
                .isInstanceOf(OptimisticLockingException.class)
                .hasMessage("Record in table 'code' has been modified prior to the delete attempt. Aborting....");
        }
    }

    @Test
    public void deleting_withExistingIdAndVersion_deletes() {
        // insert new record to the database and verify it's there
        CodeDefinition cd = new CodeDefinition("2Z", "de", new CodeClass(2, null, null), 3, true, null);
        CodeDefinition persisted = repo.saveOrUpdate(cd);
        final String code = persisted.getCode();
        final int version = persisted.getVersion();
        assertThat(repo.findCodeDefinition(code)).isNotNull();

        // delete the record
        CodeDefinition deleted = repo.delete(code, version);
        assertThat(deleted).isNotNull();
        assertThat(deleted.getCode()).isEqualTo(code);

        // verify the record is not there anymore
        assertThat(repo.findCodeDefinition(code)).isNull();
    }

    @Test
    public void gettingCodeClass1() {
        CodeClass cc1 = repo.getCodeClass1("en");
        assertThat(cc1).isNotNull();
        assertThat(cc1.getId()).isEqualTo(1);
    }

    private void assertSortedList(final String sortProperty, final String code) {
        final List<CodeDefinition> cds = repo.findPageOfCodeDefinitions(new CodeFilter(),
            new PaginationRequest(0, 10, Sort.Direction.DESC, sortProperty));

        assertThat(cds).hasSize(10);

        CodeDefinition cd = cds.get(0);
        assertThat(cd.getCode()).isEqualTo(code);
    }

    @Test
    public void findingCodeDefinitions_sortedBySort() {
        assertSortedList("sort", "1Z");
    }

    @Test
    public void findingCodeDefinitions_sortedByCode() {
        assertSortedList("code", "8Z");
    }

    private void assertFiltering(final CodeFilter filter, final int count, final String code) {
        final List<CodeDefinition> cds = repo.findPageOfCodeDefinitions(filter,
            new PaginationRequest(0, 100, Sort.Direction.DESC, "sort"));

        assertThat(cds).hasSize(count);

        CodeDefinition cd = cds.get(0);
        assertThat(cd.getCode()).isEqualTo(code);
    }

    @Test
    public void findingCodeDefinitions_filteredByInternal() {
        final CodeFilter filter = new CodeFilter();
        filter.setInternal(true);
        assertFiltering(filter, 19, "1Z");
    }

    @Test
    public void findingCodeDefinitions_filteredByCodeClass() {
        final CodeFilter filter = new CodeFilter(); // internal codeclass, comment
        filter.setCodeClass(new CodeClass(5, "whatever", ""));
        assertFiltering(filter, 13, "5U");
    }

    @Test
    public void findingCodeDefinitions_filteredByComment() {
        final CodeFilter filter = new CodeFilter(); // internal codeclass, comment
        filter.setCommentMask("COPD");
        assertFiltering(filter, 1, "4F");
    }

}

File: core/core-persistence-jooq/src/integration-test/kotlin/ch/difty/scipamato/core/persistence/JooqCodeRepoIntegrationTest.kt
package ch.difty.scipamato.core.persistence

import ch.difty.scipamato.common.entity.CodeClassId
import ch.difty.scipamato.common.logger
import ch.difty.scipamato.common.persistence.paging.PaginationRequest
import ch.difty.scipamato.common.persistence.paging.Sort
import ch.difty.scipamato.core.entity.CodeClass
import ch.difty.scipamato.core.entity.code.CodeDefinition
import ch.difty.scipamato.core.entity.code.CodeFilter
import ch.difty.scipamato.core.entity.code.CodeTranslation
import ch.difty.scipamato.core.persistence.code.JooqCodeRepo
import org.amshove.kluent.invoking
import org.amshove.kluent.shouldBeEqualTo
import org.amshove.kluent.shouldBeFalse
import org.amshove.kluent.shouldBeNull
import org.amshove.kluent.shouldBeTrue
import org.amshove.kluent.shouldContainAll
import org.amshove.kluent.shouldContainSame
import org.amshove.kluent.shouldHaveSize
import org.amshove.kluent.shouldNotBeNull
import org.amshove.kluent.shouldStartWith
import org.amshove.kluent.shouldThrow
import org.amshove.kluent.withMessage
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.fail
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.jooq.JooqTest
import org.testcontainers.junit.jupiter.Testcontainers

private val log = logger()

@JooqTest
@Testcontainers
@Suppress("TooManyFunctions", "SpellCheckingInspection", "FunctionName", "MagicNumber", "DuplicatedCode")
internal open class JooqCodeRepoIntegrationTest {

    @Autowired
    private lateinit var repo: JooqCodeRepo

    @Test
    fun findingAllCodes1InGerman() {
        val codesOfClass1 = repo.findCodesOfClass(CodeClassId.CC1, "de")
        codesOfClass1 shouldHaveSize 21
        codesOfClass1.forEach { c -> log.debug(c.toString()) }
    }

    @Test
    fun findingAllCodes2InEnglish() {
        val codesOfClass1 = repo.findCodesOfClass(CodeClassId.CC2, "en")
        codesOfClass1 shouldHaveSize 2
        codesOfClass1.forEach { c -> log.debug(c.toString()) }
    }

    @Test
    fun findingAllCodes3InEnglish() {
        val codesOfClass1 = repo.findCodesOfClass(CodeClassId.CC3, "fr")
        codesOfClass1 shouldHaveSize 14
        codesOfClass1.forEach { c -> log.debug(c.toString()) }
    }

    @Test
    fun findingCodeDefinitions_withUnspecifiedFilter_findsAllDefinitions() {
        val cds = repo.findPageOfCodeDefinitions(
            CodeFilter(),
            PaginationRequest(0, 10, Sort.Direction.ASC, "name")
        )

        cds shouldHaveSize 10

        var cd = cds[0]
        cd.code shouldBeEqualTo "4Y"
        cd.codeClass?.id shouldBeEqualTo 4
        cd.name shouldBeEqualTo "Absenzen, eingeschränkte Aktivität"
        cd.sort shouldBeEqualTo 17
        cd.isInternal.shouldBeTrue()
        cd.getNameInLanguage("de")?.shouldStartWith("Absenzen,")
        cd.getNameInLanguage("en")?.shouldStartWith("Absenteeism")
        cd.getNameInLanguage("fr")?.shouldStartWith("Absentéisme")

        cd = cds[1]
        cd.code shouldBeEqualTo "7Z"
        cd.codeClass?.id shouldBeEqualTo 7
        cd.name shouldBeEqualTo "Alle"
        cd.sort shouldBeEqualTo 4
        cd.isInternal.shouldBeTrue()
        cd.getNameInLanguage("de")?.shouldStartWith("Alle")
        cd.getNameInLanguage("en")?.shouldStartWith("All")
        cd.getNameInLanguage("fr")?.shouldStartWith("tous")
    }

    @Test
    fun findingCodeDefinitions_withFilterMatchingSingleGermanName_findsOne() {
        val filter = CodeFilter()
        filter.nameMask = "Experimentelle Studie unter Belastung / Arbeit"
        val kds = repo.findPageOfCodeDefinitions(
            filter,
            PaginationRequest(Sort.Direction.ASC, "name")
        )

        kds shouldHaveSize 1

        val ntd = kds[0]
        ntd.code shouldBeEqualTo "5A"
        ntd.name shouldBeEqualTo "Experimentelle Studie unter Belastung / Arbeit"
        ntd.getNameInLanguage("de") shouldBeEqualTo "Experimentelle Studie unter Belastung / Arbeit"
        ntd.getNameInLanguage("en") shouldBeEqualTo "Experimental study under exercising conditions"
        ntd.getNameInLanguage("fr") shouldBeEqualTo "Etude expérimentale dans des conditions exercicantes"
    }

    @Test
    fun findingCodeDefinitions_haveVersionFieldsPopulated() {
        val filter = CodeFilter()
        filter.nameMask = "Experimentelle Studie unter Belastung / Arbeit"
        val kds = repo.findPageOfCodeDefinitions(filter, PaginationRequest(Sort.Direction.ASC, "name"))

        kds shouldHaveSize 1

        val cd = kds[0]

        cd.version shouldBeEqualTo 1
        cd.created.shouldBeNull()
        cd.lastModified.shouldBeNull()

        val tr = cd.getTranslations().first()
        tr.version shouldBeEqualTo 1
        tr.created.shouldBeNull()
        tr.lastModified.shouldBeNull()
    }

    @Test
    fun countingCodes_witNullFilter_findsAllDefinitions() {
        repo.countByFilter(null) shouldBeEqualTo 82
    }

    @Test
    fun countingCodes_withUnspecifiedFilter_findsAllDefinitions() {
        repo.countByFilter(CodeFilter()) shouldBeEqualTo 82
    }

    @Test
    fun countingCodes_withFilter_findsAllMatchingDefinitions() {
        val filter = CodeFilter()
        filter.nameMask = "es"
        repo.countByFilter(filter) shouldBeEqualTo 44
    }

    @Test
    fun countingCodes_withNonMatchingFilter_findsNone() {
        val filter = CodeFilter()
        filter.nameMask = "foobar"
        repo.countByFilter(filter) shouldBeEqualTo 0
    }

    @Test
    fun gettingMainLanguage() {
        repo.mainLanguage shouldBeEqualTo "de"
    }

    @Test
    fun findingMainLanguage() {
        val cd = repo.newUnpersistedCodeDefinition()

        cd.code.shouldBeNull()
        cd.mainLanguageCode shouldBeEqualTo "de"
        cd.codeClass.shouldBeNull()
        cd.sort shouldBeEqualTo 1
        cd.isInternal.shouldBeFalse()
        cd.name shouldBeEqualTo "n.a."
        cd.getNameInLanguage("de").shouldBeNull()

        val translations = cd.getTranslations()
        translations.map { it.langCode } shouldContainSame listOf("de", "en", "fr")
        translations.map { it.id } shouldContainAll listOf(null, null, null)
        translations.map { it.name } shouldContainAll listOf(null, null, null)
    }

    @Test
    fun findingCodeDefinition_withNonExistingId_returnsNull() {
        repo.findCodeDefinition("foo").shouldBeNull()
    }

    @Test
    fun findingCodeDefinition_withExistingId_loadsWithAllLanguages() {
        val existing = repo.findCodeDefinition("4Y") ?: fail { "Unable to find code 4Y" }

        existing.code shouldBeEqualTo "4Y"
        existing.name shouldStartWith "Absenzen"
        existing.getTranslations() shouldHaveSize 3
        existing.getNameInLanguage("de")?.shouldStartWith("Absenzen,")
        existing.getNameInLanguage("en")?.shouldStartWith("Absenteeism")
        existing.getNameInLanguage("fr")?.shouldStartWith("Absentéisme")
    }

    @Test
    @Suppress("LocalVariableName", "VariableNaming")
    fun savingNewRecord_savesRecordAndRefreshesId() {
        val ct_de = CodeTranslation(null, "de", "foo_de", "Kommentar", 0)
        val ct_en = CodeTranslation(null, "en", "foo1_en", null, 0)
        val ct_fr = CodeTranslation(null, "fr", "foo1_fr", null, 0)
        val cc = CodeClass(2, "cc2", "d2")
        val cd = CodeDefinition("2Z", "de", cc, 0, false, 1, ct_de, ct_en, ct_fr)

        cd.getTranslations().map { it.id } shouldContainAll listOf(null, null, null)

        val saved = repo.saveOrUpdate(cd) ?: fail { "Unable to save code" }

        saved.code shouldBeEqualTo "2Z"
        saved.name shouldBeEqualTo "foo_de"
        saved.getTranslations().map { it.version } shouldContainAll listOf(1, 1, 1)
    }

    @Test
    fun updatingRecord() {
        val cd = repo.findCodeDefinition("2R") ?: fail { "Unable to find code 2R" }

        cd.code shouldBeEqualTo "2R"
        cd.name shouldBeEqualTo "Europa"
        cd.getTranslations() shouldHaveSize 3
        cd.getNameInLanguage("de") shouldBeEqualTo "Europa"
        cd.getNameInLanguage("en") shouldBeEqualTo "Europe"
        cd.getNameInLanguage("fr") shouldBeEqualTo "Europe"
        cd.getTranslations("de").first().version shouldBeEqualTo 1
        cd.getTranslations("en").first().version shouldBeEqualTo 1

        cd.setNameInLanguage("de", "eu")
        cd.setNameInLanguage("fr", "foo")

        val updated = repo.saveOrUpdate(cd) ?: fail { "Unable to save or update code" }

        updated.code shouldBeEqualTo "2R"
        updated.getTranslations() shouldHaveSize 3
        updated.getNameInLanguage("de") shouldBeEqualTo "eu"
        updated.getNameInLanguage("en") shouldBeEqualTo "Europe"
        updated.getNameInLanguage("fr") shouldBeEqualTo "foo"

        updated.version shouldBeEqualTo 2
        updated.getTranslations("de").first().version shouldBeEqualTo 2
        updated.getTranslations("en").first().version shouldBeEqualTo 2
        updated.getTranslations("fr").first().version shouldBeEqualTo 2
    }

    @Test
    fun deleting_withNonExistingId_returnsNull() {
        repo.delete("ZZ", 1).shouldBeNull()
    }

    @Suppress("TooGenericExceptionCaught")
    @Test
    fun deleting_withExistingId_butWrongVersion_throwsOptimisticLockingException() {
        invoking { repo.delete("1A", -1) } shouldThrow OptimisticLockingException::class withMessage
            "Record in table 'code' has been modified prior to the delete attempt. Aborting...."
    }

    @Test
    fun deleting_withExistingIdAndVersion_deletes() {
        // insert new record to the database and verify it's there
        val cd = CodeDefinition("2Z", "de", CodeClass(2, "cc2", "d2"), 3, true, null)
        val persisted = repo.saveOrUpdate(cd) ?: fail { "Unable to save or update code" }
        val code = persisted.code ?: fail { "code should not be null" }
        val version = persisted.version
        repo.findCodeDefinition(code).shouldNotBeNull()

        // delete the record
        val deleted = repo.delete(code, version) ?: fail { "Unable to delete code" }
        deleted.code shouldBeEqualTo code

        // verify the record is not there anymore
        repo.findCodeDefinition(code).shouldBeNull()
    }

    @Test
    fun gettingCodeClass1_inAKnownLanguage() {
        val cc1 = repo.getCodeClass1("en")
        cc1.id shouldBeEqualTo 1
        cc1.name shouldBeEqualTo "Exposure Agent"
    }

    @Test
    fun gettingCodeClass1_inNotKnownLanguage() {
        val cc1 = repo.getCodeClass1("foo")
        cc1.id shouldBeEqualTo 1
        cc1.name shouldBeEqualTo "not translated"
    }

    @Test
    fun findingCodeDefinitions_sortedBySort() {
        assertSortedList("sort", "1Z")
    }

    private fun assertSortedList(sortProperty: String, code: String) {
        val cds = repo.findPageOfCodeDefinitions(
            CodeFilter(),
            PaginationRequest(0, 10, Sort.Direction.DESC, sortProperty)
        )

        cds shouldHaveSize 10

        val cd = cds[0]
        cd.code shouldBeEqualTo code
    }

    @Test
    fun findingCodeDefinitions_sortedByCode() {
        assertSortedList("code", "8Z")
    }

    @Test
    fun findingCodeDefinitions_sortedByInternal() {
        val cds = repo.findPageOfCodeDefinitions(
            CodeFilter(),
            PaginationRequest(0, 10, Sort.Direction.DESC, "internal")
        )

        cds shouldHaveSize 10

        // all are internal=true, no use asserting
    }

    @Test
    fun findingCodeDefinitions_sortedByTranslation() {
        assertSortedList("translationsAsString", "2N")
    }

    @Test
    fun findingCodeDefinitions_sortedBySomethingUndefined_doesNotSort() {
        assertSortedList("notexisting", "1F")
    }

    private fun assertFiltering(filter: CodeFilter, count: Int, code: String) {
        val cds = repo.findPageOfCodeDefinitions(
            filter,
            PaginationRequest(0, 100, Sort.Direction.DESC, "sort")
        )

        cds shouldHaveSize count

        val cd = cds[0]
        cd.code shouldBeEqualTo code
    }

    @Test
    fun findingCodeDefinitions_filteredByInternal() {
        val filter = CodeFilter()
        filter.internal = true
        assertFiltering(filter, 19, "1Z")
    }

    @Test
    fun findingCodeDefinitions_filteredByCodeClass() {
        val filter = CodeFilter() // internal codeclass, comment
        filter.codeClass = CodeClass(5, "whatever", "")
        assertFiltering(filter, 13, "5U")
    }

    @Test
    fun findingCodeDefinitions_filteredByComment() {
        val filter = CodeFilter() // internal codeclass, comment
        filter.commentMask = "COPD"
        assertFiltering(filter, 1, "4F")
    }
}
